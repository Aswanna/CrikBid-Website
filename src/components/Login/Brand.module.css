.Container {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
}

.Container img {
  height: 180px;
  width: auto;
  border-radius: 10px;
}

.Container h1 {
  font-family: 'Roboto', sans-serif;
  font-weight: 500;
  font-size: 2.35rem;
  margin-bottom: -15px;
}

@media (max-width: 680px) {
  .Container {
    flex-direction: row;
  }
  .Container img {
    width: 100px;
    height: auto;
    margin-right: 15px;
  }
  .Container h1 {
    font-size: 1.75rem;
  }
  .Container span {
    font-size: 0.75rem;
  }
}

@media (max-width: 300px) {
  .Container h1 {
    font-size: 1.4rem;
  }

  .Container img {
    width: 75px;
    height: auto;
  }

  .Container span {
    font-size: 0.6rem;
  }
}
