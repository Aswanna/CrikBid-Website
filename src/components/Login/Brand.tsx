import React from 'react';
import Logo from '../../assets/icons/me-logo.png';
import classes from './Brand.module.css';

const Brand = () => {
  return (
    <div className={classes.Container}>
      <img src={Logo} alt="brand-logo" />
      <div>
        <h1>Match eka</h1>
        <span>Powered by SeyComp.Lk</span>
      </div>
    </div>
  );
};

export default Brand;
