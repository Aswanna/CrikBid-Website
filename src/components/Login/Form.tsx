import React from 'react';
import Input from '../UI/Input';
import { InputProps } from '../../types';

const Form = ({ email, pwd }: { email: InputProps; pwd: InputProps }) => {
  return (
    <>
      <Input {...email} />
      <Input {...pwd} />
    </>
  );
};

export default Form;
