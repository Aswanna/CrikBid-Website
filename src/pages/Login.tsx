import React, { FormEventHandler, useContext, useState } from 'react';
import useInput from '../hooks/useInput';
import Brand from '../components/Login/Brand';
import Form from '../components/Login/Form';
import Button from '../components/UI/Button';
import classes from '../styles/Login.module.css';
import { InputProps } from '../types';
import { authContext } from '../context/authContext';
import { Link } from 'react-router-dom';

const Login = () => {
  const emailValidator = useInput(
    (inputVal) => inputVal.includes('@') && !inputVal.includes(' ')
  );
  const pwdValidator = useInput((inputVal) => inputVal.trim().length > 6);
  const [message, setMessage] = useState<string | null>(null);
  const validatorArr: { email: InputProps; pwd: InputProps } = {
    email: {
      type: 'text',
      value: emailValidator.inputValue,
      onChange: emailValidator.valueChangeHandler,
      onBlur: emailValidator.inputBlurHandler,
      placeholder: 'Enter your email',
      hasError: emailValidator.hasError,
      errorMessage: 'Enter a valid email address',
      ref: emailValidator.inputRef,
      label: undefined,
      style: {},
    },
    pwd: {
      type: 'text',
      value: pwdValidator.inputValue,
      onChange: pwdValidator.valueChangeHandler,
      onBlur: pwdValidator.inputBlurHandler,
      placeholder: 'Enter your password',
      hasError: pwdValidator.hasError,
      errorMessage: 'Enter a valid password',
      ref: pwdValidator.inputRef,
      label: undefined,
      style: {},
    },
  };

  const ctx = useContext(authContext);

  const submitFormHandler: FormEventHandler<HTMLFormElement> = async (e) => {
    e.preventDefault();

    if (!emailValidator.isInputValid) return emailValidator.focusHandler();

    if (!pwdValidator.isInputValid) return pwdValidator.focusHandler();

    const body = {
      email: emailValidator.inputValue,
      password: pwdValidator.inputValue,
    };

    const url = `${process.env.REACT_APP_API_ENDPOINT}user/login`;
    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify(body),
      headers: {
        'Content-Type': 'application/json',
      },
    });
    const data = await response.json();
    if (data.success) {
      ctx.login(data.token, false, data.expiresOn);
    } else {
      setMessage(data.message);
    }
  };
  return (
    <>
      <div className={classes.NavCont}>About</div>
      <div className={classes.Container}>
        <div className={classes.BrandBar}>
          <Brand />
        </div>
        <div className={classes.MiddleBar}></div>
        <form className={classes.FormBar} onSubmit={submitFormHandler}>
          <div className={classes.Topic}>
            <h1>Login</h1>
            <p>Welcome to match eka!</p>
          </div>
          <Form email={validatorArr.email} pwd={validatorArr.pwd} />
          {message && <div className="error">{message}</div>}
          <Button type="submit" style={undefined} data-testid="submit-button">
            <span>Login</span>
          </Button>

          <div className={classes.CreateCont}>
            <h1>Don't have an account?</h1>
            <Link to="/user/register">Create an account</Link>
          </div>
        </form>
      </div>
    </>
  );
};

export default Login;
