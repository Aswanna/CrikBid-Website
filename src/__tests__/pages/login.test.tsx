import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import Login from '../../pages/Login';

afterAll(() => {
  cleanup();
});

describe('Intergration tests of /login', () => {
  test('Should match snapshot', () => {
    const LoginPage = render(<Login isAdmin={false} />, {
      wrapper: MemoryRouter,
    });
    expect(LoginPage).toMatchSnapshot();
  });

  test('Should show an error when email is invalid after typing', () => {
    const { queryByPlaceholderText, queryByText } = render(
      <Login isAdmin={false} />,
      {
        wrapper: MemoryRouter,
      }
    );
    const emailInput = queryByPlaceholderText('Enter your email') as Element;
    expect(queryByText('Enter a valid email address')).not.toBeInTheDocument();
    fireEvent.change(emailInput, {
      target: {
        value: 'user',
      },
    });

    fireEvent.blur(emailInput);
    expect(queryByText('Enter a valid email address')).toBeInTheDocument();
  });

  test('Should show an error when password is invalid after typing', () => {
    const { queryByPlaceholderText, queryByText } = render(
      <Login isAdmin={false} />,
      {
        wrapper: MemoryRouter,
      }
    );
    const pwdInput = queryByPlaceholderText('Enter your password') as Element;
    expect(queryByText('Enter a valid password')).not.toBeInTheDocument();
    fireEvent.change(pwdInput, {
      target: {
        value: 'pwd',
      },
    });

    fireEvent.blur(pwdInput);
    expect(queryByText('Enter a valid password')).toBeInTheDocument();
  });

  test('Should show an error in email if button is clicked without entering a single char', () => {
    const { queryByRole, queryByText } = render(<Login isAdmin={false} />, {
      wrapper: MemoryRouter,
    });
    const button = queryByRole('button') as Element;
    expect(queryByText('Enter a valid email address')).not.toBeInTheDocument();
    fireEvent.click(button);
    expect(queryByText('Enter a valid email address')).toBeInTheDocument();
  });
});
