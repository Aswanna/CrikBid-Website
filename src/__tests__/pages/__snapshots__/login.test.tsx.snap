// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Intergration tests of /login Should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="NavCont"
      >
        About
      </div>
      <div
        class="Container"
      >
        <div
          class="BrandBar"
        >
          <div
            class="Container"
          >
            <img
              alt="brand-logo"
              src="me-logo.png"
            />
            <div>
              <h1>
                Match eka
              </h1>
              <span>
                Powered by SeyComp.Lk
              </span>
            </div>
          </div>
        </div>
        <div
          class="MiddleBar"
        />
        <form
          class="FormBar"
        >
          <div
            class="Topic"
          >
            <h1>
              Login
            </h1>
            <p>
              Welcome to match eka!
            </p>
          </div>
          <div
            class="Container"
          >
            <input
              class="Input"
              placeholder="Enter your email"
              type="text"
              value=""
            />
          </div>
          <div
            class="Container"
          >
            <input
              class="Input"
              placeholder="Enter your password"
              type="text"
              value=""
            />
          </div>
          <button
            class="Button"
            type="submit"
          >
            <span>
              Login
            </span>
          </button>
          <div
            class="CreateCont"
          >
            <h1>
              Don't have an account?
            </h1>
            <a
              href="/user/register"
            >
              Create an account
            </a>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="NavCont"
    >
      About
    </div>
    <div
      class="Container"
    >
      <div
        class="BrandBar"
      >
        <div
          class="Container"
        >
          <img
            alt="brand-logo"
            src="me-logo.png"
          />
          <div>
            <h1>
              Match eka
            </h1>
            <span>
              Powered by SeyComp.Lk
            </span>
          </div>
        </div>
      </div>
      <div
        class="MiddleBar"
      />
      <form
        class="FormBar"
      >
        <div
          class="Topic"
        >
          <h1>
            Login
          </h1>
          <p>
            Welcome to match eka!
          </p>
        </div>
        <div
          class="Container"
        >
          <input
            class="Input"
            placeholder="Enter your email"
            type="text"
            value=""
          />
        </div>
        <div
          class="Container"
        >
          <input
            class="Input"
            placeholder="Enter your password"
            type="text"
            value=""
          />
        </div>
        <button
          class="Button"
          type="submit"
        >
          <span>
            Login
          </span>
        </button>
        <div
          class="CreateCont"
        >
          <h1>
            Don't have an account?
          </h1>
          <a
            href="/user/register"
          >
            Create an account
          </a>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
